planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,397 @@
+      }
+    }
+    return false;
+  }
+
+  public LoopLineMerger setPrecisionModel(PrecisionModel precisionModel) {
+    this.precisionModel = precisionModel;
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public void add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+  }
+
+  private void merge() {
+    for (var node : output) {
+      if (node.getEdges().size() == 2) {
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
+      }
+    }
+  }
+
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
+        if (allPaths.size() > 1) {
+          for (var path : allPaths.subList(1, allPaths.size())) {
+            for (var toRemove : path) {
+              toRemove.remove();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  List<List<Edge>> findAllPaths(Node start, Node end, double maxLength) {
+    List<List<Edge>> allPaths = new ArrayList<>();
+    Queue<List<Edge>> queue = new LinkedList<>();
+
+    for (var edge : start.getEdges()) {
+      List<Edge> path = new ArrayList<>();
+      path.add(edge);
+      if (maxLength > 0.0) {
+        if (edge.length <= maxLength) {
+          queue.add(path);
+        }
+      } else {
+        queue.add(path);
+      }
+    }
+
+    while (!queue.isEmpty()) {
+      List<Edge> currentPath = queue.poll();
+      Node currentPoint = currentPath.getLast().to;
+
+      if (currentPoint == end) {
+        allPaths.add(new ArrayList<>(currentPath));
+      } else {
+        for (var edge : currentPoint.getEdges()) {
+          if (!currentPath.contains(edge)) {
+            List<Edge> newPath = new ArrayList<>(currentPath);
+            newPath.add(edge);
+            if (maxLength > 0.0) {
+              if (getLength(newPath) <= maxLength) {
+                queue.add(newPath);
+              }
+            } else {
+              queue.add(newPath);
+            }
+          }
+        }
+      }
+    }
+
+    allPaths.removeIf(LoopLineMerger::hasPointAppearingMoreThanTwice);
```

Comment:
Consider the following graph:


And say you want to get from A to F without visiting an edge twice. The solutions for this are:

- ABCEF
- ABCEBDEF
- ABCEDBEF
- ABDEF
- ABDEBCEF
- ABDECBEF
- ABEF
- ABECBDEF
- ABEDBCEF

But I don't want to keep these sort of spiraling paths so that is why I remove them here. It might well be that with a loopMinLength of 20 as you have it in one of the tests, this function goes into crazy branching and that is why id does not finish, maybe.

Reply:
I moved the check now to before the `queue.add(newPath)` call